Add option to generate new index using updated unexpected results in compare-static-analysis-results

briannafan opened the pull request:
#### f283632e7ab8857dd6bd8f70d3f64889588756c2
<pre>
Add option to generate new index using updated unexpected results in compare-static-analysis-results
<a href="https://bugs.webkit.org/show_bug.cgi?id=282713">https://bugs.webkit.org/show_bug.cgi?id=282713</a>
<a href="https://rdar.apple.com/139383076">rdar://139383076</a>

Reviewed by Ryosuke Niwa.

Add --generate-results-only option. This will be used in Safer-CPP-Checks after comparing the unexpected
results to results database. If there are changes in the results, then we need to generate a new results index,
without performing another comparison.

Usage: compare-static-analysis-results [new-dir] --build-output [scan-build-output] --scan-build-path [scan-build] --generate-results-only

* Tools/Scripts/compare-static-analysis-results:
(parser): Add generate-results-only option.
(compare_results): Separate out comparison logic from main.
(generate_results_only): Read from unexpected_results.json and issues_per_file.json to get paths to the unexpected result files.
    Pass those files into the pre-existing create_filtered_results_dir function to generate an index with new results.
(main): Clean up main.

Canonical link: <a href="https://commits.webkit.org/286309@main">https://commits.webkit.org/286309@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae2644f286099339e98de818b0072fdbf6e91940

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75567 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/28418 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80047 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/26831 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77683 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/2815 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/59274 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/26831 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78634 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/28418 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39632 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/28418 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25159 "Built successfully") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/28418 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81526 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/2906 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/2815 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67513 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/3057 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/28418 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66810 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/28418 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11667 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/2863 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/5685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/2888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/3823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/2895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Tools/Scripts/compare-static-analysis-results:
```diff
@@ -332,8 +341,35 @@ def main():
+    return results_for_upload
+
+
+def generate_results_only(args):
```
I'd call this generate_filtered_results(args) or something.

rniwa commented on Tools/Scripts/compare-static-analysis-results:
```diff
@@ -332,8 +341,35 @@ def main():
+def main():
+    args = parser()
+
+    if not args.generate_results:
```
It's confusing that this variable is called "generate results". We're always generating results.
Maybe generate_filtered_results?

rniwa commented on Tools/Scripts/compare-static-analysis-results:
```diff
@@ -332,8 +341,35 @@ def main():
+    for project, checkers in results_data.get('failures').items():
+        for checker, files in checkers.items():
+            for file in files:
+                result_paths += [i for i in issues_per_file[checker][file].values()]
```
issues_per_file.values() don't contain integers does it? It's confusing to call that variable "i".
If it's really an index, why is the aggregated result called "result_paths"?

briannafan replied:
No, sorry, I used `i` as short for `issue`.

I think better naming is `report_paths`, since these are actually paths to the reports for each issue (e.g. `scan-build-output/StaticAnalyzer/WebCore/StaticAnalyzerReports/report-ContainerNode.cpp-parserInsertBefore-13-a66ea5.html`). We need to pass these into `create_filtered_results_dir` to create a filtered `StaticAnalyzerReports` folder that `scan-build --generate-index-only` uses.

I've made these changes, let me know what you think.

briannafan committed:
Add option to generate new index using updated unexpected results in compare-static-analysis-results
https://bugs.webkit.org/show_bug.cgi?id=282713
rdar://139383076

Reviewed by Ryosuke Niwa.

Add --generate-results-only option. This will be used in Safer-CPP-Checks after comparing the unexpected
results to results database. If there are changes in the results, then we need to generate a new results index,
without performing another comparison.

Usage: compare-static-analysis-results [new-dir] --build-output [scan-build-output] --scan-build-path [scan-build] --generate-results-only

* Tools/Scripts/compare-static-analysis-results:
(parser): Add generate-results-only option.
(compare_results): Separate out comparison logic from main.
(generate_results_only): Read from unexpected_results.json and issues_per_file.json to get paths to the unexpected result files.
    Pass those files into the pre-existing create_filtered_results_dir function to generate an index with new results.
(main): Clean up main.

Canonical link: https://commits.webkit.org/286309@main
